docs/spec/reactors/consensus/proposer-selection.md

```diff
@@ -2,45 +2,268 @@
+
+If we consider the validator set where p3 has just been added:
+
+Validator | p1 | p2 | p3
+----------|--- |--- |---
+VP        | 1  | 3  | 8
+
+Let's assume that the last round was R and the proposer priorities were as shown in first row (R end) with their sum being 0. p3 is added with A(p3) = -4 (penalty loss due to integer division?)
+In the next round, p3 will still be ahead in the queue, elected as proposer and move back in the queue (same position it was just added)
+
+|Round Priority  | -4 | -3 | -2 | -1 | 0  | 1 | 2 | 3 | 4 | Alg step
+|--------------- | ---|--- |----|--- |--- |---|---|---|---|--------
+|H/R end         |    |    | p2 |    |    |   | p1|   |   |add p3
+|                | p3 |    | p2 |    |    |   | p1|   |   |A(p3) = -4
+|H+1/0           |    |    |    |    |    | p2|   | p1| p3|A(i)+=VP(i)
+|                | p3 |    |    |    |    | p2|   | p1|   |A(p3)-=P
+
+### Proposer Priority Range
+With the introduction of shifting, an interesting case occurs. Low power validators that bind early benefit from subsequent addition of high power validators. This is because new validators are added with negative priorities and this causes an increase in priority for all other validators during the shifting operation. As a consequence, low power validators may end up with high priority and election algorithm will fail with respect to fairness.
+
+As an example, let's assume the following set, where p2 has just been added in the first row (H/R end)
+
+Validator | p1| p2 
+----------|---| ---
+VP        | 10| 100
+
+The proposer selection runs as described in the New Validator section. Notice the priority of p1 bumping from 10 to 15 because of the shifting. Then in the last row shown the power of p2 is changed to 1 at the end of H+1. 
+
+|Round Priority| -21 | -16 | -15 |-11 | 0  | 10 | 15 | 25 | 89 |  Comment
+|--------------| --- |---- |---- |--- |--- |--- |--- |--- |--- |---------
+| H/R end      |     |     |     | p2 | p1 |    |    |    |    | added p2
+| H+1/0        |     |     |     |    |    | p1 |    |    | p2 | A(i)+=VP(i)
+|              | p2  |     |     |    |    | p1 |    |    |    | A(p2)-= P, P=110
+|              |     | p2  |     |    |    |    | p1 |    |    | A(i) -= avg, avg=-5
+| VP(p2)<-1    |     |     | p2  |    |    |    |    | p1 |    | A(i)+=VP(i)
+...
+
+In this example p2 happened to be at the end of the queue when its power changed to 1 and it will now crawl for ~40 rounds to catch up with p1 even though their power ratio is 1:10.
+
+In order to prevent these types of cases, the selection algorithm ensures that the range of priority values is not bigger than two times the total voting power. 
+
+The modified selection algorithm is:
+
+    def singleProposerSelection (vset):
+        for each validator i in vset:
+            A(i) += VP(i)
+        prop = max(A)
+        A(prop) -= P
+    
+    def ProposerSelection (vset, times):
+        // normalize the priority values
+	    if max(A)-min(A) > 2 * P:
```

Comment:
It seems that this algorithm incentives validator that is at the end of the queue to decrease it's voting power (by delegating to another validator it owns) so he can move ahead in the queue, and then to increase its voting power. The same as above, this step should be executed upon validator set changes only and not during normal proposer selection.

Reply:
Heh - https://github.com/cosmos/cosmos-sdk/tree/develop/docs/spec/staking

Basically the voting power of a validator in "Bonded Proof of Stake" systems like Cosmos is proportional to the amount of tokens "bonded" to that validator, where bonding can be done by anyone with tokens. People who bond their tokens to a validator are called "delegators"